### PR TITLE
Define Strict CSP in the Authoring Considerations section.

### LINKED ISSUE #528: Add Strict CSP in Authoring Considerations
Strict CSP is a well known CSP configuration against XSS which is both effective and deployable. Therefore, we should add a section for Strict CSP in Authoring Considerations.

This has been discussed in a WebAppSec mail thread.
https://lists.w3.org/Archives/Public/public-webappsec/2021Oct/0007.html

### PR DESCRIPTION
**Pull request by shhnjk**
Fixes #528.

### CODE REVIEW

**Comment by shhnjk** [comment]
@mikewest, PTAL! BTW, how can I see changes in actual visual (i.e. is there a way to build this locally?)?

File: index.bs [hunk 1]
```diff
@@ -674,10 +686,10 @@ spec: WebAssembly-web-api-api; urlPrefix: https://webassembly.github.io/spec/web
                   ; <a>path-absolute</a> is defined in section 3.3 of RFC 3986.
 
     ; Keywords:
-    <dfn>keyword-source</dfn> = "<dfn>'self'</dfn>" / "<dfn>'unsafe-inline'</dfn>" / "<dfn>'unsafe-eval'</dfn>"
-                     / "<dfn>'strict-dynamic'</dfn>" / "<dfn>'unsafe-hashes'</dfn>" /
-                     / "<dfn>'report-sample'</dfn>" / "<dfn>'unsafe-allow-redirects'</dfn>"
-                     / "<dfn>'wasm-unsafe-eval'</dfn>"
+    <dfn>keyword-source</dfn> = "<dfn>'none'</dfn>" / "<dfn>'self'</dfn>" / "<dfn>'unsafe-inline'</dfn>"
```
Comment (mikewest) [inline index.bs:689]: This change (as well as the change on 671 above) looks unrelated to what you're actually trying to do in this patch. Perhaps move it to another PR for discussion?

File: index.bs [hunk 2]
```diff
@@ -4964,6 +4976,41 @@ this algorithm returns normally if compilation is allowed, and throws a
...
+        and/or <em>hash</em> <a grammar>source-expression</a> with the
+        "<a grammar>'strict-dynamic'</a>" <a grammar>keyword-source</a>.
+    2.  <em>base-uri</em>: "<a grammar>'self'</a>" or "<a grammar>'none'</a>"
+        <a grammar>keyword-source</a>s.
+    
+    A CSP that meets the above criteria is called Strict CSP. Further details
+    are discussed in [[WebDev-StrictCSP]].
+
+    <div class="example">
+      Followings are few examples of Strict CSP.
```
Comment (mikewest) [inline index.bs:5001]:
```suggestion
      The following are examples of Strict CSP:
```

File: index.bs [hunk 3]
```diff
@@ -4964,6 +4976,41 @@ this algorithm returns normally if compilation is allowed, and throws a
...
+
+    Deployment of an effective CSP against XSS is a challenge (as described in
+    <a href="https://dl.acm.org/doi/10.1145/2976749.2978363">CSP Is Dead, Long
+    Live CSP!</a> [[LONG-LIVE-CSP]]). However, enforcing the following set of CSP
+    directives has been identified as an effective and deployable mitigation
+    against XSS.
+
+    1.  <em>script-src</em>: only use <em>nonce</em> <a grammar>source-expression</a>
+        and/or <em>hash</em> <a grammar>source-expression</a> with the
+        "<a grammar>'strict-dynamic'</a>" <a grammar>keyword-source</a>.
```
Comment (mikewest) [inline index.bs:4993]: 1. Is it reasonable to add a discussion of backwards compatibility requirements, given that Safari still hasn't shipped with `strict-dynamic`?
2. It might also be reasonable to point to `strict-dynamic` as something to avoid if possible.
3. It might also also be reasonable to point to the union of one strict policy and another policy that confines script to a given set of origins as being even more robust.

**Review by mikewest — COMMENTED** [review]
This looks like a good start to me. Perhaps @lweichselbaum can weigh in?

File: index.bs (hunk 1 above)
Comment (shhnjk) [reply index.bs:689]: I have added the following line in the Strict CSP section:

```
base-uri: "'self'" or "'none'" keyword-sources.
```

But the spec defines `'none'` as _serialized-source-list_ even though `'self'` is a _keyword-source_. If I explain that way, the line above becomes confusing.

Any better spec way of stating both `'none'` and `'self'`?

Comment (mikewest) [reply index.bs:689]: The spec suggests that a CSP directive should have _either_ a serialized source list, _or_ "`'none'`". I'm happy to believe that that was a mistake, but it's one we would want to correct in a distinct PR.

For this patch, I think you can just drop "keyword-sources" from the sentence you point to. Something like `<em>base-uri</em>: Specify a value of either "<a grammar>'self'</a>" or "<a grammar>'none'</a>".` sounds fine to me.

Comment (shhnjk) [reply index.bs]: Done :)

File: index.bs (hunk 3 above)
Comment (shhnjk) [reply index.bs:4993]: Addressed 1 and 2 with the latest commit. For 3, I'm okay adding it, but not sure how much we should expand on "This is much better" solutions. 

IMO, those are better documented in the "CSP deployment Wiki"™️, and link to the Wiki should be added here eventually.

Let we know Mike and @lweichselbaum 😊

**Comment by lweichselbaum** [comment]
I already worked with @shhnjk on the draft, so this looks good to me.

**Review by mikewest — APPROVED** [review]
If Lukas is happy, I'm happy. I'll land this.
